Wizard deploy documentation (#415)
OpenZeppelin/defender-docs: docs/modules/ROOT/pages/wizard-plugin.adoc

[[contracts-wizard-deploy-plugin]]
= Contracts Wizard Deploy Plugin

When configuring contracts from https://wizard.openzeppelin.com/[Contracts Wizard, window=_blank], you can directly deploy the configured Smart Contract using your Defender account.

[[usage]]
== Usage

[[api-key-generation]]
=== API Key generation
In your Defender dashboard, go to *Settings -> API Keys* and click *Create API Key*, you only need _Manage Deployments_ permission.

WARNING: We also recommend to set an expiration for the API Key, considering that is going to be used from an external site.

image::remix-plugin-api-key.png[Defender Remix Plugin Api Key, 400, 400]

[[deploying-from-wizard]]
=== Deployment from Contracts Wizard

Go to https://wizard.openzeppelin.com/[Contracts Wizard, window=_blank] site, and after editing your contract, click on "Deploy with Defender".

> TODO add image here

Reviewer: We have TODO here

Reviewer 2: Beat me to it :P

Author: > We have TODO here

I'm was trying to engage open source contributors to do it for me 😉

[[configure]]
==== Configure
Set your *API Key* and *API Secret* and press "Authenticate". You should see a message below the button indicating that the credentials are valid.

image::wizard-plugin-configure.png[Defender Wizard Plugin Configure]
image::wizard-plugin-configure-2.png[Defender Wizard Plugin API keys]

[[network]]
==== Network

> Select any of the supported networks. This also includes private and fork networks configured in your tenant.

Reviewer 2: ```suggestion
Select any of the supported networks. This also includes private and forked networks configured in your tenant.
```

image::wizard-plugin-network-2.png[Defender Wizard Plugin Network]

[[approval-process]]
==== Approval Process
Here you have 3 options:

- Select an existing approval process from your *Deployment Environment* configured for the selected network.

NOTE: If you have an existing deployment environment in the selected network, this is the only option allowed.

- If the *Deployment Envoronment* does not exist for the selected network, then you can create a new one. 

NOTE: If the Approval Process to be created is a Relayer, the API Key must include _Manage Relayers_ permission.

- Additionally, you can use the *injected provider* from Remix (a browser wallet) to deploy the contract, this will create a Defender *Deployment Environment* under the hood after deploying the contract.

image::wizard-plugin-approval-process.png[Defender Wizard Plugin Approval Process]

[[deploy]]
==== Deploy

> In this step, you should see the constructor inputs of your configured contract (if any). And the option to create a deterministic deployment.

Reviewer 2: ```suggestion
In this step, you should see the constructor inputs of your configured contract (if any), and the option to create a deterministic deployment.
```

> NOTE: This step is reactive, if you modify the contract you will see the new constructor arguments updated after a second.

Reviewer 2: ```suggestion
NOTE: This step is reactive, if you modify the contract you will see the new constructor arguments updated right after.
```

WARNING: Upgradable contracts are not yet fully supported. This action will only deploy the implementation contract without initializing. For safe upgrades, we strongly recommend usign https://github.com/OpenZeppelin/openzeppelin-upgrades[Upgrades Package, window=_blank].

image::wizard-plugin-deploy.png[Defender Wizard Plugin Deploy]

[[deterministic-deployments]]
==== Deterministic Deployments

> Defender Deploy supports a `salt` value to create deployments to deterministic addresses using `create2`. Click on `Deterministic` checkbox and set the salt field to any arbitrary value.

Reviewer 2: ```suggestion
Defender Deploy supports a `salt` value to create deployments to deterministic addresses using `CREATE2`. Click on `Deterministic` checkbox and set the salt field to any arbitrary value.
```

NOTE: If the approval process selected is a Multisig, the `salt` is required as Defender only support deterministic deployments when using Multisigs.

image::wizard-plugin-deterministic.png[Defender Wizard Plugin Deterministic Deployments]

[[further-steps]]
==== Further Steps

> Once the contract deployment was submitted to Defender, in some cases you will need to complete the deployment from Defender Dashboard, you should see message indicating that the contract was submitted and a button that redirects to your Deployment in Defender.

Reviewer 2: ```suggestion
Once the contract deployment is submitted to Defender, in some cases you may need to complete the deployment from Defender Dashboard. You will see message indicating that the contract was submitted and a button that redirects to your Deployment in Defender.
```

Author: Thanks, I need to retake those English grammar classes 🤦‍♂️
